File: src/main/java/org/cerion/stocklist/web/CombinedDataAPI.java
package org.cerion.stocklist.web;

import org.cerion.stocklist.Price;
import org.cerion.stocklist.model.Dividend;
import org.cerion.stocklist.model.Interval;
import org.cerion.stocklist.model.Quote;
import org.cerion.stocklist.model.Symbol;
import org.cerion.stocklist.web.api.GoogleFinance;
import org.cerion.stocklist.web.api.Tiingo;
import org.cerion.stocklist.web.api.YahooFinance;

import java.util.Date;
import java.util.List;
import java.util.Map;
import java.util.Set;

public class CombinedDataAPI implements DataAPI {

    private YahooFinance yahoo;
    private GoogleFinance google;
    private Tiingo tiingo;

    public CombinedDataAPI() {
        yahoo = YahooFinance.Companion.getInstance();
        google = new GoogleFinance();
        tiingo = new Tiingo();
    }

    @Override
    public List<Price> getPrices(String symbol, Interval interval, Date start) throws Exception {
        return yahoo.getPrices(symbol, interval, start);
    }

    @Override
    public List<Dividend> getDividends(String symbol) {
        return yahoo.getDividends(symbol);
    }

    @Override
    public List<Symbol> getSymbols(Set<String> symbols) {
        throw new UnsupportedOperationException();
    }

    @Override
    public Symbol getSymbol(String symbol) {
        try {
            return tiingo.getSymbol(symbol);
        }
        catch (Exception ignored) {

        }

        return null;
    }

    @Override
    public Map<String, Quote> getQuotes(Set<String> symbols) {
        throw new UnsupportedOperationException();
    }

    @Override
    public Quote getQuote(String symbol) {
        return google.getQuote(symbol);
    }
}

File: src/main/kotlin/web/CombinedDataAPI.kt
package org.cerion.stocks.core.web

import org.cerion.stocks.core.PriceList
import org.cerion.stocks.core.PriceRow
import org.cerion.stocks.core.model.Dividend
import org.cerion.stocks.core.model.Interval
import org.cerion.stocks.core.model.Quote
import org.cerion.stocks.core.model.Symbol
import org.cerion.stocks.core.web.clients.*
import java.util.*

class CombinedDataAPI : DataAPI {

    private val yahoo: YahooFinance = YahooFinance.instance
    private val google: GoogleFinance = GoogleFinance()
    private val tiingo: Tiingo = Tiingo()

    override fun getPriceList(symbol: String, interval: Interval, start: Date): PriceList {
        val prices = getPrices(symbol, interval, start)
        val list = PriceList(symbol, prices)
        list.lastUpdated = Date()

        return list
    }

    @Throws(Exception::class)
    override fun getPrices(symbol: String, interval: Interval, start: Date): List<PriceRow> {
        return yahoo.getPrices(symbol, interval, start)
    }

    override fun getDividends(symbol: String): List<Dividend> {
        return yahoo.getDividends(symbol)
    }

    override fun getSymbols(symbols: Set<String>): List<Symbol> {
        throw UnsupportedOperationException()
    }

    override fun getSymbol(symbol: String): Symbol? {
        try {
            return tiingo.getSymbol(symbol)
        } catch (ignored: Exception) {

        }

        return null
    }

    override fun getQuotes(symbols: Set<String>): Map<String, Quote> {
        throw UnsupportedOperationException()
    }

    override fun getQuote(symbol: String): Quote? {
        return google.getQuote(symbol)
    }
}
